pkg/controller/nodeconfig/nodeconfig_controller.go

```diff
@@ -0,0 +1,462 @@
+	return f(name)
+}
+
+type jobOwnerControl struct {
+	controller *Controller
+}
+
+func (c *jobOwnerControl) GetOwnerKey(owner metav1.Object) (string, error) {
+	return controller.KeyFunc(owner)
+}
+
+func (c *jobOwnerControl) GetOwner(namespace string, name string) (metav1.Object, error) {
+	return c.controller.machineSetsLister.MachineSets(namespace).Get(name)
+}
+
+func (c *jobOwnerControl) OnOwnedJobEvent(owner metav1.Object) {
+	machineSet, ok := owner.(*clusteroperator.MachineSet)
+	if !ok {
+		c.controller.logger.WithFields(log.Fields{"owner": owner.GetName(), "namespace": owner.GetNamespace()}).
+			Errorf("attempt to enqueue owner that is not a machineset")
+		return
+	}
+	c.controller.enqueueMachineSet(machineSet)
+}
+
+type jobSyncStrategy struct {
+	controller *Controller
+}
+
+func (s *jobSyncStrategy) GetOwner(key string) (metav1.Object, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if len(namespace) == 0 || len(name) == 0 {
+		return nil, fmt.Errorf("invalid key %q: either namespace or name is missing", key)
+	}
+	return s.controller.machineSetsLister.MachineSets(namespace).Get(name)
+}
+
+func (s *jobSyncStrategy) DoesOwnerNeedProcessing(owner metav1.Object) bool {
+	machineSet, ok := owner.(*clusteroperator.MachineSet)
+	if !ok {
+		s.controller.logger.Warn("could not convert owner from JobSync into a machineset: %#v", owner)
+		return false
+	}
+
+	msLog := logging.WithMachineSet(s.controller.logger, machineSet).WithField("func", "DoesOwnerNeedProcessing")
+	if !machineSet.Status.Installed {
+		msLog.Debug("control plane not installed, returning false")
+		return false
+	}
+
+	// Process if the machine set generation has changed since we last ran:
+	if machineSet.Status.NodeConfigInstalledJobGeneration != machineSet.Generation {
+		msLog.WithFields(log.Fields{
+			"NodeConfigInstalledJobGeneration": machineSet.Status.NodeConfigInstalledJobGeneration,
+			"Generation":                       machineSet.Generation,
+		}).Debug("machine set generation has changed")
+		return true
+	}
+	return false
+}
+
+func (s *jobSyncStrategy) DoesOwnerNeedReProcessing(owner metav1.Object) bool {
+	machineSet, ok := owner.(*clusteroperator.MachineSet)
+	if !ok {
+		s.controller.logger.Warn("could not convert owner from JobSync into a machineset: %#v", owner)
+		return false
+	}
+
+	msLog := logging.WithMachineSet(s.controller.logger, machineSet).WithField("func", "DoesOwnerNeedReProcessing")
+
+	// Process if it has been more than the configurable interval of time since we last ran:
+	installingCondition := controller.FindMachineSetCondition(machineSet, clusteroperator.MachineSetNodeConfigInstalling)
+	installedCondition := controller.FindMachineSetCondition(machineSet, clusteroperator.MachineSetNodeConfigInstalled)
+	if installedCondition == nil || installingCondition == nil {
+		msLog.Warn("machine set has node config installed generation but missing relevant conditions")
```

Comment:
why? it could be that it's just not ready for processing yet (ie. master not installed)